### ISSUE TITLE
Async reader may return wrong results on MacOS, Linux, WSL, Docker

### ISSUE THREAD

**Issue by Nercury** [dotnet/SqlClient#659]
### Describe the bug

Sometimes the query may return wrong results under high load, on Non-Windows clients.

### To reproduce

- .NET Core
- MSSQL instance
- Client is not running on Windows (reproduced on MacOS or Linux)
- Query uses async reader methods
- High load that saturates the connection pool and causes timeouts.

Bug is seen more often when `MultipleActiveResultSets=True` (MARS) is in the connection string. It might be prevented by disabling MARS (unclear at this point).

The issue is hard to reproduce and rare, therefore [this project was created with the code that attempts to simulate the situation](https://github.com/Nercury/mars-wrong-results).

This program starts 2000 concurrent connections that runs `select @Id as Id`
statement. Each `Id` is different, and the query result should always return 
the id that was queried, yet under certain circumstances that's not the case. It looks like
sometimes the reader returns the stale data from random previous connection that experienced the timeout.

There are more details in the linked project.

Network latency can be more easily simulated using this method here: https://github.com/trainline/SqlClient659

### Expected behavior

Under no circumstances a select statement should return a different result. 

### Further technical details

Microsoft.Data.SqlClient version: 4.8.1 or earlier
.NET target: netcoreapp2.0 - netcoreapp3.1
SQL Server version: SQL Server Web Edition 13.00.5426.0.v1, SQL Server Express Edition 14.00.3281.6.v1
Operating system: macOS 10.15.5, Docker container

### Workaround

Do not use this SQL client on linux/docker.

Always check the id of the returned result and retry or crash if it is wrong.

To decrease the likelihood of this or maybe prevent it (unclear at this point), do not use MARS feature, if you are using .NET Core with MSSQL database, ensure that the connection string does not contain `MultipleActiveResultSets=True`.

**Comment by JRahnama** [comment]
@Nercury thanks for reaching out. Issue #422 is similar to the problem you have and is under investigation.

**Comment by DavoudEshtehari** [comment]
Hi @Nercury,

Thank you for producing the sample app and describing the issue well.
I ran your sample app multiple times at a virtual machine running `Ubuntu 18.0`, `.Net core 3.1`, and `SQL Server 19` with `MARS` enabled and `connection pool` equals **100**. I didn't get any `Wrong results` as you expected.

Do you have any comments?

**Comment by Nercury** [comment]
> I didn't get any `Wrong results` as you expected.
> 
> Do you have any comments?

Yes, this matches my experience. As I mentioned, the issue seems to be related to timeouts, and they can not happen if the DB server and the code is on the same machine, and the query is doing basically nothing. 

Try to introduce some latency between the db server and the application. 

I reproduced it using a bit crude approach, by running DB server in cloud and the application on my laptop with a router in another room (so there is some considerable latency between the two). Maybe there are tools to do it a bit more cleanly.

In production environment the AWS starts throttling read operations when they go above allowed IOPS limit, that's how it originally happened when running on AWS.

**Comment by DavoudEshtehari** [comment]
By replacing the data source by a slower connection over the network I am able to reproduce the wrong result.
I checked the code and added a `local index` inside the `QueryAndVerify` method to keep the `index` parameter value and used this local variable instead of that parameter inside the function body and this change solved the issue. Can I ask you to give it a try and let me know?

**Comment by Nercury** [comment]
Thanks for checking it out. I am not sure I completely understand what you mean by "added a local index inside the QueryAndVerify method to keep the index parameter value and used this local variable instead of that parameter inside the function body", because the `int index` is already local to the `QueryAndVerify` function. Can you provide a changeset (in the form of PR) so that I am sure I am doing the same thing?

**Comment by DavoudEshtehari** [comment]
While I was preparing the requested changes I found out it is happening again. Maybe the network conditions simultaneously had affected my test by the last change.
The issue intermittently happens either with `System.Data.Sqlclient` & `Microsoft.Data.SqlClient` when `ExecuteReaderAsync` blends with exceptions.

**Comment by DavoudEshtehari** [comment]
In between investigation, I found these mitigations besides a workaround that has mentioned as disabling the `MARS`:

- Set **Pooling = false**
- Removing a connection that encountered an exception from the connection pool: **`SqlConnection.ClearPool(cn)`**
- Attempting to execute a SQL command without attaching a transaction to the command (remove the `BeginTransaction` & `CommitAsync`) :
> `var command = new SqlCommand(sql, cn);`
> `command.Transaction = null;`

- Using **`ExecuteScalarAsync`** instead of `ExecuteReaderAsync`

@Nercury Can you give them a try one by one?

**Comment by Nercury** [comment]
## Set Pooling = false

`MultipleActiveResultSets=True;Pooling=false;` painfully slow and many timeouts, but no wrong results, which is good

## Removing the connection that encountered an exception

```
            SqlConnection cnp = null;
            try
            {
                using (var cn = new SqlConnection(connectionString))
                {
                    cnp = cn;
                    
                    // ...
                }
            }
            catch (Exception e)
            {
                if (cnp != null)
                {
                    SqlConnection.ClearPool(cnp);
                }
                
                // ...
            }
```
Got wrong results on a third attempt.

## Attempting to execute a SQL command without attaching a transaction to the command

This was attempted in production when it was still unclear if this is a MARS issue. It makes wrong results more rare, but they still happen. I was not able to reproduce it in this test project with a simplistic select statement.

## Using ExecuteScalarAsync instead of ExecuteReaderAsync

Tried it 5 times and wrong results did not happen.

@DavoudEshtehari I hope this helps.

**Comment by gyphysics** [comment]
### Same bug experienced when MARS is disabled:

We are experiencing a similar problem in our code base, although we have MARS disabled explicitly in the connection string and still see the bug.

We run this method across multiple incoming requests from an API Controller:

```
public async Task<T> GetAsync(string id)
        {
            var sql = $"SELECT Data FROM { _configuration.TableName } {(_configuration.InMemory ? "WITH (SNAPSHOT)" : "")} WHERE Id=@id";

            using (var connection = _connectionFactory.Create())
            {
                var record = await connection.QueryAsync<string>(sql, new { id }).ConfigureAwait(false);

                var data = record.FirstOrDefault();

                return data == null ? default(T) : _serializer.Deserialize<T>(data);
            }
        }
```
The connectionFactory simply returns a new SqlConnection:
`return new SqlConnection(_connectionString);`

`id` is a GUID
`data` will be an encrypted blob of data that decrypts to a JSON payload.
`_serializer` handles the decryption (if needed) then serialisation.

Things will run normally, then a single instance will experience a SQL Exception - previously there was an unexpected (and denied) request to perform a SHOWPLAN, sometimes it's a SQL Timeout exception.  Following that exception, serialisation bugs will continue to appear on that instance until it is taken out of service.  

The source of the serialisation failure is that the value stored in `data` is an `id` from an earlier request, rather than the expected value stored on the SQL DB.  We've had this happen both on the in-memory DB and without it.

Once the instance has been out of service for a while, subsequent attempts to access the DB via this method complete successfully, so it seems to have the ability to resolve itself.  Even when the instance is misbehaving, it appears that not all requests get jumbled up, as some manage to succeed.

Previously this service was hosted on Windows without any issues.  Since the port to Linux this problem crops up every 7-30 days or so.  We've also seen it on more than one service (the above code is in a shared library).  This bug seems to be the closest thing to what we are experiencing, except that we have explicitly disabled MARS in the connection string and are still seeing it from time to time.

**Libraries/Infrastructure involved:**
NetCore 3.1
Dapper 2.0.35
Microsoft.Data.SqlClient 2.0.1

This is being run on an AWS T3 instance hosting Ubuntu 18.04

We are going to attempt changing the call from QueryAsync to ExecuteScalarAsync and are attempting to capture some more information to help us debug the issue.  If the change to ExecuteScalarAsync resolves the problem for at least a month or so then I'll add an extra comment here indicating that it's a possible fix.

**Comment by cheenamalhotra** [comment]
Hi @Nercury @gyphysics 

Could you please try your use-cases with PR #796 as I tried to run repro provided and am able to see issue fixed.
Kindly confirm.

**Comment by Nercury** [comment]
> This issue remained unfixed for now, due to colliding changes in different async paths.

@cheenamalhotra All right, it makes sense to wait until all fixes are in.

**Comment by simonjduff** [comment]
Following on from @gyphysics (some company, different team):
Upgraded to 2.1.0 and still seeing this issue in multiple services, all on Linux.
Definitely no MARS in connection string (though we haven't actively disabled it).
We are using Dapper and will try to remove it from one project to prove it's not related.
In the meantime we're using guard clauses to prevent wrong data being used
```
if (!thingId.Equals(thing.Id))
{
    throw new ContextConflictException($"Asked for thing [{thingId}], got thing [{thing.Id}]");
}
```

We have also seen some problems on INSERT, where no exception was raised, but checking later the data was not written to the table. There were gaps in the IDENTITY column, suggesting a possible silent transaction rollback (exception didn't come back to originating async context?). Possibly unrelated, but seems close enough to link not here.

**Comment by simonjduff** [comment]
I have just set up a test case using @Nercury 's test program, deliberately creating network latency towards the sql server using iptables
```
sudo iptables -A INPUT -p tcp --sport 1433 -m statistic --mode random --probability 0.1 -j DROP
```
This is with a connection string with no MARS, and I was able to get a wrong result.
Connectionstring:
```
Connection string Server=SERVERADDRESS;Database=DATABASE;Integrated Security=False;User Id=USERNAME;Password=PASSWORD!;MultiSubnetFailover=true;Max Pool Size=100
```

**Comment by alefranz** [comment]
Hello!

I've created a new repro putting together items of this discussion using `docker-compose` so it can be easily execute locally also simulating network failures.

I can confirm the problem can be reproduced even without MARS, although it is less frequent.

| .NET version | SQL Server Version | MARS | Failures rate |
| ------------ | ------------------ | ---- | ------- |
| .NET Core 3.1 | 2017 Enterprise | Yes | High |
| .NET Core 3.1 | 2017 Enterprise | No | Low |
| .NET 5.0 | 2017 Enterprise | Yes | High |
| .NET 5.0 | 2017 Enterprise | No | Low |
| .NET Core 3.1 | 2019 Enterprise | Yes | High |
| .NET Core 3.1 | 2019 Enterprise | No | Low |
| .NET 5.0 | 2019 Enterprise | Yes | High |
| .NET 5.0 | 2019 Enterprise | No | Low |

@cheenamalhotra Any suggestion on how to try to pinpoint the origin of the problem? Is there a diagnostic source available in the SqlClient?

**Comment by mgravell** [comment]
Hi; I see a mention of Dapper here. It is entirely possible that this is a Dapper issue, using an unintended and unsupported feature that only kinda works by accident (and only when MARS is enabled). Ultimately, "MARS enabled" is not the same as "performing evilness via overlapped unawaited operations", and I will be the first to say: any time you don't immediately await something, you're *effectively* doing the same thing as multi-threading, and race conditions become obvious.

So; if there's a problem, and this API doesn't work (or even claim to work) the way in which Dapper has used (abused) it: maybe we should just deprecate that feature from Dapper - or at least see if we can re-work it to be less evil.

(I'm one of the Dapper maintainers)

**Comment by Nercury** [comment]
The system where the bug was discovered was using Dapper too. But the bug remained after the code was rewritten without Dapper. We will know for sure once the underlying issue is fixed.

**Comment by alefranz** [comment]
@Nercury is it worth updating the issue title as it is not only happening with MARS enabled?

**Comment by alefranz** [comment]
Both repros were tracking together queries returning invalid result with queries returning no result (but no exception).

I've now updated https://github.com/trainline/SqlClient659 to track those separately.
Without MARS I am not able to get invalid results with this repro so I will try to create a new one.

However not returning records where it should and not throwing an exception is a bug as well.

**Comment by Nercury** [comment]
@alefranz let me know how it goes, because I have removed the MARS as a possible workaround now based on your previous report. I should add it back if not using MARS does not return wrong result. Returning no results is not that bad (but not good either).

**Comment by cheenamalhotra** [comment]
Posting latest update for investigations so far:

For scenario:
![image](https://user-images.githubusercontent.com/13396919/104404322-41b23300-550f-11eb-955b-f118953770cd.png)

This is what is happening:
![image](https://user-images.githubusercontent.com/13396919/104404372-5a224d80-550f-11eb-8ec6-a8b0e660a39f.png)

We are receiving Data on pooled connection AFTER connection is sent back to pool, and that data gets added to Read Packet Queue in SNI Mars Handle, from where it's picked up in next read cycle for next query, leading to "wrong data".

Lack of effective tracing is making the investigations slow, but we're working on improving and adding more meaningful traces to entire driver that can help track end-to-end connection activity in complex applications.

Thanks for your patience!

**Comment by Wraith2** [comment]
@cheenamalhotra continuing here from out discussion of this on an unrelated PR

> We should be cleaning up MarsHandle or packets which are still connected to StateObj on the connection in pooling.
I also always hit the Debug assert in SNIPacket's destructor, and we have a lot of unreleased packets, maybe it's related?

Hitting the finalizer on SNIPackets regularly is worrying. At the top of the file there's this comment:
```csharp
// #define TRACE_HISTORY // this is used for advanced debugging when you need to trace the entire lifetime of a single packet, be very careful with it
```

try enabling that and see if you can replicate the problem, if you can then you'll be able to look at the history of the packet to identify what allocated it and then with some work you can trace back to the thing that should have, but didn't, release it. 

Every managed packet should be allocated and released through the pool and the pool forces the data to be cleared by calling Release() on it. If the packet is dropped to the GC it can't be recycled and the array rented from the pool would be lost as well so it shouldn't be possible for a dropped packet to be re-used.

The packet could be partially reconstructed while the disconnect happens on another thread. The packet smux header contains the connection id used to find the connection to dispatch it to. There's a dictionary of SNIMarsHandle in SNIMarsConnection (iirc) and that's where the dispatch happens so at that point the connection must have the correct handler in the dictionary to find it and send the packet to it. If there's a queue on the mars handle then that should be cleared, though really as I understand the TDS protocol the server won't send any packets after an end message so it'd be weird for there to be any pending.

**Comment by cheenamalhotra** [comment]
Here's a typical example of what happens:

![image](https://user-images.githubusercontent.com/13396919/104537687-48a07a80-55cf-11eb-94bd-bff71439cd78.png)
[Left column is time in milliseconds]
If you see carefully they have data left of 67 bytes - which is not consumed or requested before. The 2 times above 54 and 468 bytes were read and consumed from this packet.

For every case where we get "Execution Timeout expired" errors, we have these finalizers getting hit.
![image](https://user-images.githubusercontent.com/13396919/104538075-0cb9e500-55d0-11eb-80a0-c70ce2435f37.png)

These unreleased packets also lead to packet Ids reaching 250/300 sometimes, I donno what is the max we can handle if max pool size is 100 (as default).

**Comment by Wraith2** [comment]
If there is no pooled packet available a new packet is created, similarly if the pool is full and one is returned it is dropped to the GC. Not being able to get or return a packet to the pool is not a problem and is inefficient but not an error. The only time a packet getting to GC is a problem when it has a payload in it which is why there's the null check in the finalizer, getting finalized with data means something failed to return a live packet to the pool which doesn't cause an error but also shouldn't happen because it makes the pool ineffective.

The packet id is per-process so seeing it get up to 300 really isn't too bad. The default packet pool per physical connection is 4 which is the maximum number that a non-mars connection can have in use at any one time as far as I can tell. 1 Send, 1 Cancel, 1 Received, 1 Receiving. A densely used mars connection multiplexed on top of the physical connection could easily need more packets than that.

If you want to entirely eliminate the pool just comment out the pool TryGet and Return calls in SNIPhysicalHandle and change the array pool allocation to a simple new[] call, that'll means no packets are recycled. If the error stops occurring then it was connected to recycling, if it doesn't you know you're safe to ignore the pool and look elsewhere.

I'm very curious about what those 67 remaining bytes are in the dropped packets. Is it possible to grab the entire packet buffer on one of them and see if we can determine what is being taken and what the 67 bytes left are?

**Comment by cheenamalhotra** [comment]
Hi Everyone!

Providing this week's progress so far:

We have managed to capture and study all detailed traces for this behavior, including Event Source, SQL Server Profiler, and Wireshark to strip open full story which I'll detail below:

**Fact we know:**
When an exception occurs on Client side, client shall send an Attention packet to the server to immediately stop processing further data and to not send any further data. On the client side, sending attention can mean something too, as for example ensuring transaction is rollbacked and data at rest is cleared properly or whether connection should be killed and not reused. It depends on when and how we receive an exception and send attention.

**Problem we see here:**
We have 2 situations going on when exception "Execution Timeout Expired" is to be received, and it looking like there's a race condition between Transaction Commit and Attention when Read timeout occurs in SNIMarsHandle.
- **[GOOD FLOW]** Begin Trans > Run Query > Sends Attention > Trans Rollback
  - Connection throws exception and terminates safely.
- **[BAD FLOW]** Begin Trans > Run Query > Trans Commit > Sends Attention
  - Query succeeds in this case as Commit succeeded but when connection is reused from the pool, the next query throws "Execution Timeout Expired" exception followed by a rollback but data is kept in store and not cleared up > leading to wrong data flow.

We know this is related to client side timeout that occurs during Read as Read operations are blocked due to thread starvation, so even though server sends data on time, client does not read it on time. But, even if we manage to fix the thread starvation issue, that is not the solution to this particular issue, as the BAD FLOW is the real problem and can occur in any similar situation.

Will provide more updates as we continue to study our traces to dig deeper and find a solution.

**Comment by cheenamalhotra** [comment]
@Wraith2 

I will come back to packet recycling and cleaning again, it may not be related here, but we will fix that too eventually.

Let's try and find possible race condition scenarios in Commit and Attention signals when timeout exception occurs in SNIMarsHandle, as it feels we might be able to restrict the BAD FLOW by ensuring we also wait for Error handling to complete and only pass data to user if there are no exceptions as here we have data as well as exception coming in from timeout. We don't want exception to flow down to next connection.

**Comment by cheenamalhotra** [comment]
Hi everyone!

Thank you for your patience, we seem to have a fix available to try.
Would request you to please try out the PR https://github.com/dotnet/SqlClient/pull/894 and let us know!

[Download Build CI NuGet](https://dev.azure.com/sqlclientdrivers-ci/sqlclient/_build/results?buildId=23336&view=artifacts&pathAsName=false&type=publishedArtifacts)

**Comment by alefranz** [comment]
Hi Cheena,
great news! thanks for the update!
Is this fix on a codepath used only when MARS is enabled? or could this also fix the harder ro reproduce incorrect behaviour without MARS?

**Comment by cheenamalhotra** [comment]
This fix applies to SqlCommand layer and to Begin and End methods of Execute Reader Async so Mars was not a factor, and neither is Managed SNI.

We also know ExecuteXmlReaderAsync also has same issue, I will apply the same fix there on Monday.

**Comment by alefranz** [comment]
That's great news, I was really struggling to find a good repro for the non-MARS scenario! I will start testing this build next week.

**Comment by cheenamalhotra** [comment]
I now have a consistent repro of this issue to rule out inconsistencies, as it's always been difficult to consistently reproduce this. The issue is exaggerated with Threadpool exhaustion and instead of exhausting resources everytime, we are able to put a small delay on one of the delegates we know are causing problem here and are able to reproduce the wrong data use case.

[Link to Repro](https://gist.github.com/cheenamalhotra/5f4b35603c491484e51be38abf15548a)

To run this repro, you'd need to reference driver source project and update [OnTimeout](https://github.com/dotnet/SqlClient/blob/cde615e83612635d397c290ce033a617c7c67280/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs#L2250) function with a small delay period of 10 seconds, so it looks like this:

```cs
private void OnTimeout(object state)
    {
       Thread.Sleep(10000); // Include this line
       if (!_internalTimeout)
       {
           ...
```

The output captured is as under:

![image](https://user-images.githubusercontent.com/13396919/107474760-cc467c00-6b27-11eb-9e3b-eeb4c33247b1.png)

While we're trying our best to fix this by various methods, things are very flaky. and design is highly complex. Being able to reproduce consistently does give some hope. But there may be more cases where wrong data could be possible. Will provide more update as we identify more scenarios and solve this one.

**Comment by cheenamalhotra** [comment]
Hi Everyone!

We'd like to request community validations on the fix from PR #906.
We've captured the root cause and fixed the issue in the core parser layer.

Please give the below NuGet package a try and send us your feedback:
[Microsoft.Data.SqlClient.2.1.0-pull-89d6662.21043.7](https://dev.azure.com/sqlclientdrivers-ci/_apis/resources/Containers/8693000/Nuget?itemPath=Nuget%2FMicrosoft.Data.SqlClient.2.1.0-pull-89d6662.21043.7.nupkg)

**Comment by Wraith2** [comment]
Anyone?

**Comment by simonjduff** [comment]
We're very keen @Wraith2, just need to find a moment having spent engineering time moving linux services back to windows to mitigate the issue.

**Comment by patrickbouwens1978** [comment]
@Wraith2 , I've spend some time testing the pull request with my own repro-case.

Earlier on, I was able to get the incorrect results with the general version 2.1.1. Consistent wrong results over the different attempts.

I've also tried Microsoft.Data.SqlClient.2.1.0-pull-3bdbb71.21048.9. This version never gave any incorrect results, so it looks like this fix solves the problem.

**Comment by alefranz** [comment]
Hi @cheenamalhotra , @Wraith2 ,

as you know reproducing the issue, especially without MARS, on real world applications is extremely challenging, so it is hard to prove the fix.
I can confirm however that it doesn't introduce any appreciable regression.

Thanks for the great effort on this!

**Comment by patrickbouwens1978** [comment]
Some more background information about my earlier repro-case, for reference.

The reprocase is ran in Docker in a Microsoft Azure environment, on a Linux App Service Plan (P1V2) and SQL Server Azure (S0)
The base dockerfile is mcr.microsoft.com/dotnet/core/aspnet:3.1.8-bionic, so non-windows environment on .Net core 3.1
Connectionstring didn’t include any MARS-references, so all is used in a non-MARS environment.

The base situation is a full async operation, so openasync, executereaderasync, and readasync. I've never used transactions and a low commandtimeout (1 second) to enforce problems as much as possible..

The SQL Statement I used is “select @Id as Id”

The issue of the wrong results only occurred when a “Timeout expired exception” appeared. Scaling up / down the database also causes exceptions, but didn’t trigger any wrong results (or at least, I couldn’t manage to do so).

**The results:**
Base scenario: use of openasync, executereaderasync, and readasync, high load situation with multiple threads, multiple requests.
When things went wrong, they kept going wrong in a consistent matter. For instance, suppose there are three users:
User 1: Select * FROM User where UserId = 1
User 2: Select * FROM User where UserId = 2
User 3: Select * FROM User where UserId = 3
User 1 caused an exception. User 2 received the results from User 1, User 3 received the results from User 2, and so on.
Different requests got each other’s results, even after some time. More and more SQL statements results in wrong results, even after some time. The connection pool become corrupted, and stayed corrupted.
The only solution was to restart the web app.

**Mitigations:**
I tried a number of mitigations.

_Attempt 1: Use ConfigureAwait(false) for openasync, executereaderasync, readasync._
No effect, wrong results keep appearing like in the base scenario.

_Attempt 2: Use .Open() instead of OpenAsync()_
Exception are occurring more often (due to Timeout expired), but only a few wrong results.  Furthermore, the wrong results weren’t consistly wrong. New requests could be ok, so no restart of the application was required. So it looked like some connetions became corrupt, but somehow, it didn't influence the connectionpool for later connections.

Performance was, however, reduced significly.

**Results with the patch:**
In all situation, no wrong results have occurred.

**In summary:**
The only mitigation I could find, was to:
* Reduce load, so the chance of Timeout Expired exception was reduced
* Use Open instead of OpenAsync, so the chance of wrong results was reduced as well, and wrong results could ‘disappear’ in future requests. I’ve had some luck using an O/R mapper, so I could change to use Open instead of OpenAsync with just a couple of lines of code.

The patch solves the situation of wrong results appearing. So looking forward for this hotfix to be released... ;)

**Comment by Wraith2** [comment]
The underlying bug that was found was caused by threadpool exhaustion. It is a bug in the TdsParser layer which is above the network layer so it was possible to cause the problem in any version of the driver but was much more likely in managed (linux, macos) mode and with mars due to the way they use threads. 

The issue itself was down to timing of internal cancellation messages and how they were implemented which made it possible for a timeout to arrive after a result had returned, or even worse after another query had begun. What the PR does is harden the timeout handling by adding synchronous timeout checks and carefully tracking the origin of the async timeout so that if it doesn't match the executing query it is discarded. This results in more accurate timeouts in high contention scenarios and prevention of cross query timeouts which were causing a strange mismatch in the driver.

**Comment by cheenamalhotra** [comment]
The latest Microsoft.Data.SqlClient [v2.1.2](https://www.nuget.org/packages/Microsoft.Data.SqlClient/2.1.2) has been released containing fix for this issue.
Closing as resolved.

**Comment by alefranz** [comment]
Thank you all for the brilliant work to track down this elusive bug!

**Comment by andrehaupt** [comment]
@cheenamalhotra / @Wraith2: We are experiencing a similar issue using versions 1.1.x. Is this issue applicable to v2.1 only?

**Comment by Wraith2** [comment]
It could occur in all versions of SqlClient prior to the fix in 2.1.1

**Comment by andrehaupt** [comment]
@Wraith2 Thank you.

**Comment by cheenamalhotra** [comment]
Hi @andrehaupt 

[v1.1.4](https://www.nuget.org/packages/Microsoft.Data.SqlClient/1.1.4) has been released to contain the fix too. Please upgrade if you're still using the v1.1 series.

**Comment by tlecomte** [comment]
Hi @cheenamalhotra, @Wraith2,

Would you be able to clarify the following points please ?

- was this bug also affecting .NET 5 apps running on Windows ? The title of this issue seems to say that it should not happen on Windows, but looking at the fix in #906 it seems like it also affects Windows, and even .NET Framework apps, isn't it ?

- is System.Data.SqlClient also affected - again in the context of a .NET 5 app running on Windows ?

I'm trying to analyze a similar issue observed with a .NET 5 app running on Windows with System.Data.SqlClient (and .NET runtime 5.0.2), so these clarifications would greatly help, thanks!

**Comment by Wraith2** [comment]
> was this bug also affecting .NET 5 apps running on Windows ? 

The problem could affect all versions of the library running on any version of .net. It was more likely on the managed driver but the root cause was shared. It was a timing issue.

> is System.Data.SqlClient also affected

Yes, every version of SqlClient prior to the fix could be affected. 
It doesn't look like the fix was backported to corefx so the System version retains the ability to encounter the problem https://github.com/dotnet/corefx/pulls?q=is%3Apr+is%3Aclosed+label%3Aarea-System.Data.SqlClient I don't know whether servicing was considered but given the rarity of the problem and the difficulty of the fix i can see why it wouldn't be be an obvious choice to push back to an LTS release.
